### PR TITLE
Implement IOClient

### PR DESCRIPTION
**Pull request by firstdoit**
#### What is the purpose of this pull request?

We decided to deprecate IODataSource a long time ago. The time has come.

#### What problem is this solving?

DataSources are a **Terrible Idea®**.

#### Types of changes

* [ ] Bug fix (a non-breaking change which fixes an issue)
* [ ] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: src/HttpClient/HttpClient.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    const postConfig: RequestConfig = {...config, url, data, method: 'post'}
+    return this.request(postConfig) as Promise<IOResponse<T>>
+  }
+
+  public patch = <T = void>(url: string, data?: any, config: RequestConfig = {}): Promise<T> => {
+    const patchConfig: RequestConfig = {...config, url, data, method: 'patch'}
+    return this.request(patchConfig).then(response => response.data as T)
+  }
+
+  public delete = (url: string, config?: RequestConfig): Promise<IOResponse<void>> => {
```
Comment (marcosvcp) [inline src/HttpClient/HttpClient.ts]: Why delete does not receive a generic type definition like others? I think it should

Comment (firstdoit) [reply src/HttpClient/HttpClient.ts]: Delete shouldn't have a response, right? Anyway, didn't change this bit.

Comment (marcosvcp) [reply src/HttpClient/HttpClient.ts]: Nope, delete should include a deleted-entity if response was 200

Comment (kevinch) [reply src/HttpClient/HttpClient.ts]: Agreed with Marcos, i think it should return the deleted element.

File: src/HttpClient/AppClient.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { IOContext } from '../service/typings'
+
+import { IOClient } from './IOClient'
+import { AuthType, InstanceOptions } from './typings'
+
+/**
+ * Used to perform calls on apps you declared a dependency for in your manifest.
+ */
+export class AppClient extends IOClient {
```
[the remark this answers is not in the document]
Comment (firstdoit) [reply src/HttpClient/AppClient.ts]: Wow, complicated question. I'm not sure. One option is to have a separate CrossAccountAppClient. Another option is to indicate in the params...

File: src/HttpClient/AppClient.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  constructor(app: string, context: IOContext, options: InstanceOptions) {
```
[the remark this answers is not in the document]
Comment (firstdoit) [reply src/HttpClient/AppClient.ts]: Disagree. You should only create requests for apps you depend, and you can only depend on one version. Hence, `app` is necessarily `vendor.name`.

File: src/IODataSource.ts [hunk 4]
```diff
@@ -10,6 +10,9 @@ interface HttpClientFactoryOptions {
 
 export type HttpClientFactory = (opts: HttpClientFactoryOptions) => HttpClient | void
 
+/**
+ * @deprecated Extend one of the subclasses of HttpClient instead.
```
Comment (tlgimenes) [inline src/IODataSource.ts]:
```suggestion
 * @deprecated Extend one of the subclasses of IOClient instead.
```

File: src/clients/LicenseManager.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      metric: 'lm-account-data',
+    })
+  }
+
+  public getTopbarData (VtexIdclientAutCookie: string) {
+    return this.http.get(routes.topbarData, {
+      headers: {
+        VtexIdclientAutCookie,
+      },
+      memoizeable: false,
```
Comment (tlgimenes) [inline src/clients/LicenseManager.ts]:
```suggestion
      memoizeable: true,
```

**Pull request merged**